src/input/mod.rs

```diff
@@ -1574,6 +1510,227 @@ impl State {
+                                let res = shell.move_request(
+                                    &surface,
+                                    &seat_clone,
+                                    serial,
+                                    ReleaseMode::NoMouseButtons,
+                                    false,
+                                    &state.common.config,
+                                    &state.common.event_loop_handle,
+                                    &state.common.xdg_activation_state,
+                                );
+                                if let Some((target, focus)) = res {
+                                    seat_clone.modifiers_shortcut_queue().clear();
+
+                                    std::mem::drop(shell);
+                                    if target.is_touch_grab() {
+                                        seat_clone
+                                            .get_touch()
+                                            .unwrap()
+                                            .set_grab(state, target, serial);
+                                    } else {
+                                        seat_clone
+                                            .get_pointer()
+                                            .unwrap()
+                                            .set_grab(state, target, serial, focus);
+                                    }
+                                }
+                            });
+                        }
+                    } else {
+                        under = Some(target);
+                    }
+                } else {
+                    let layers = layer_map_for_output(&output);
+                    if let Some(layer) = layers
+                        .layer_under(WlrLayer::Bottom, relative_pos.as_logical())
+                        .or_else(|| {
+                            layers.layer_under(WlrLayer::Background, relative_pos.as_logical())
+                        })
+                    {
+                        let layer_loc = layers.layer_geometry(layer).unwrap().loc;
+                        if layer.can_receive_keyboard_focus()
+                            && layer
+                                .surface_under(
+                                    relative_pos.as_logical() - layer_loc.to_f64(),
+                                    WindowSurfaceType::ALL,
+                                )
+                                .is_some()
+                        {
+                            under = Some(layer.clone().into());
+                        }
+                    };
+                }
+            }
+        }
+        std::mem::drop(shell);
+        Shell::set_focus(self, under.as_ref(), seat, Some(serial), false);
+    }
+
+    /// Derives a keyboard focus target from a global position. This is a variant of maybe_update_focus
+    /// that doesn't require write lock on the shell.
```

Comment:
The focus-update code gets increasingly complicated and I am not a big fan of this, given we duplicate a bunch of that logic now. From what I am seeing, this doesn't need `Common`/`self`, but only the shell.

Couldn't this be a method of the shell instead, that we could also use inside `maybe_update_focus`?

Reply:
I moved it the shell, and use it inside of `maybe_update_focus`. I wasn't sure if I should leave the modifier key check in maybe update, given it was checked immediately prior to that in `keyboard_focus_from_position` function. 

I could see an argument for keeping it or leaving it.